presto-hive/src/main/java/com/facebook/presto/hive/materializedview/MaterializedViewUtils.java

```diff
@@ -0,0 +1,254 @@
+    {
+        List<Column> partitionColumns = table.getPartitionColumns();
+        List<String> partitionNamesInOrder = metastore.getPartitionNames(table.getDatabaseName(), table.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(new SchemaTableName(table.getDatabaseName(), table.getTableName())));
+
+        List<Map<String, String>> partitionSpecs = new ArrayList<>();
+        for (String partitionName : partitionNamesInOrder) {
+            Map<String, String> partitionSpec = toPartitionSpec(partitionName);
+            if (partitionColumns.size() != partitionSpec.size()) {
+                throw new PrestoException(HIVE_INVALID_METADATA, String.format("Expected %d partition key values, but got %d", partitionColumns.size(), partitionSpec.size()));
+            }
+            for (Column partitionColumn : partitionColumns) {
+                String name = partitionColumn.getName();
+                HiveType hiveType = partitionColumn.getType();
+                if (!hiveType.isSupportedType()) {
+                    throw new PrestoException(NOT_SUPPORTED, String.format("Unsupported Hive type %s found in partition keys of table %s.%s", hiveType, table.getDatabaseName(), table.getTableName()));
+                }
+                if (partitionSpec.get(name) == null) {
+                    throw new PrestoException(HIVE_INVALID_PARTITION_VALUE, String.format("partition key value cannot be null for field: %s", name));
+                }
+                // partitionSpec.replace(name, HIVE_DEFAULT_DYNAMIC_PARTITION, "\\N");
+            }
+            partitionSpecs.add(partitionSpec);
+        }
+
+        List<String> partitionColumnNames = partitionColumns.stream().map(Column::getName).collect(toList());
+        return new MaterializedViewPredicatesInfo(partitionSpecs, partitionColumnNames);
+    }
+
+    public static MaterializedViewPredicatesInfo diffPartitionSpecs(MaterializedViewPredicatesInfo leftSpecs, MaterializedViewPredicatesInfo rightSpecs, Map<String, String> rightToLeftPartitionKeyMap)
+    {
+        if (rightToLeftPartitionKeyMap.isEmpty()) {
+            return emptyMaterializedViewPredicatesInfo();
+        }
+        if (rightSpecs.isEmpty()) {
+            return leftSpecs;
+        }
+
+        Set<String> leftMappedCommonKeys = new HashSet<>();
+        Set<String> rightMappedCommonKeys = new HashSet<>();
+        for (String rightKey : rightSpecs.getColumnNames()) {
+            String leftKey = rightToLeftPartitionKeyMap.get(rightKey);
+            if (leftKey != null && leftSpecs.getColumnNames().contains(leftKey)) {
+                leftMappedCommonKeys.add(leftKey);
+                rightMappedCommonKeys.add(rightKey);
+            }
+        }
+        if (leftMappedCommonKeys.isEmpty()) {
+            return emptyMaterializedViewPredicatesInfo();
+        }
+
+        Set<Map<String, String>> rightSpecsMappedToLeftCommonKeys = new HashSet<>();
```

Comment:
call it `commonPredicates`; also, using a `Set` to dedup a map could be risky. It depends on what map implementation we are using; it may not achieve the goal of finding the same map in terms of the content.